File: app/core/pcapwriter.py
import struct
import time
from typing import List, Dict, Any


def save_pcap_file(filepath: str, packets: List[Dict[str, Any]]) -> bool:
    """
    Saves packet records into a standard libpcap binary file (.pcap / .pcapng).
    Excludes AbuseIPDB, VirusTotal, and IPinfo threat intel attributes, preserving
    100% native compatibility with Wireshark and tshark.
    """
    if not filepath:
        return False

    # Standard libpcap global header (24 bytes)
    # Magic: 0xa1b2c3d4 (little-endian), Major: 2, Minor: 4, Thiszone: 0, Sigfigs: 0, Snaplen: 65535, Network: 1 (Ethernet)
    global_header = struct.pack("<IHHiIII", 0xa1b2c3d4, 2, 4, 0, 0, 65535, 1)

    try:
        with open(filepath, "wb") as f:
            f.write(global_header)
            base_time = time.time()

            for idx, pkt in enumerate(packets):
                raw_bytes = pkt.get("raw_bytes", b"")
                if not raw_bytes:
                    # Construct minimal Ethernet frame if raw_bytes missing
                    raw_bytes = b"\x00\x11\x22\x33\x44\x55\x66\x77\x88\x99\xaa\xbb\x08\x00"

                # Extract or calculate packet timestamp
                time_str = pkt.get("time", "")
                ts_sec = int(base_time) + idx
                ts_usec = 0

                if time_str:
                    try:
                        parts = time_str.split(".")
                        time_struct = time.strptime(parts[0], "%H:%M:%S")
                        # Combine current date with packet HH:MM:SS
                        now_struct = time.localtime()
                        full_time = time.struct_time((
                            now_struct.tm_year, now_struct.tm_mon, now_struct.tm_mday,
                            time_struct.tm_hour, time_struct.tm_min, time_struct.tm_sec,
                            now_struct.tm_wday, now_struct.tm_yday, now_struct.tm_isdst
                        ))
                        ts_sec = int(time.mktime(full_time))
                        if len(parts) > 1:
                            ts_usec = int(parts[1][:3]) * 1000
                    except Exception:
                        pass

                length = len(raw_bytes)
                # Packet record header (16 bytes): ts_sec, ts_usec, incl_len, orig_len
                pkt_header = struct.pack("<IIII", ts_sec, ts_usec, length, length)
                f.write(pkt_header)
                f.write(raw_bytes)

        return True
    except Exception as e:
        print(f"[PCAPWriter] Error writing PCAP file {filepath}: {e}")
        return False

File: SentinelSharkDotNet/Core/PcapWriter.cs
using System;
using System.Collections.Generic;
using System.IO;
using SentinelSharkDotNet.Models;

namespace SentinelSharkDotNet.Core;

public static class PcapWriter
{
    public static bool SavePacketsAuto(string filepath, List<PacketInfo> packets)
    {
        if (filepath.EndsWith(".pcapng", StringComparison.OrdinalIgnoreCase))
        {
            try
            {
                PcapNgWriter.SavePackets(filepath, packets);
                return true;
            }
            catch
            {
                return false;
            }
        }
        return SavePcapFile(filepath, packets);
    }

    public static bool SavePcapFile(string filepath, List<PacketInfo> packets)
    {
        try
        {
            using var fs = new FileStream(filepath, FileMode.Create);
            using var bw = new BinaryWriter(fs);
            
            // Write global header (24 bytes)
            bw.Write(0xa1b2c3d4u); // magic
            bw.Write((ushort)2);    // version major
            bw.Write((ushort)4);    // version minor  
            bw.Write(0);            // thiszone
            bw.Write(0u);           // sigfigs
            bw.Write(65535u);       // snaplen
            bw.Write(1u);           // network (Ethernet)
            
            foreach (var pkt in packets)
            {
                byte[] data = pkt.RawBytes ?? BuildEthernetFrame(pkt);
                
                uint tsSec = 0;
                uint tsUsec = 0;
                
                if (DateTime.TryParse(pkt.Time, out DateTime ts))
                {
                    tsSec = (uint)((DateTimeOffset)ts).ToUnixTimeSeconds();
                    tsUsec = (uint)(ts.Millisecond * 1000 + ts.Microsecond);
                }
                else
                {
                    tsSec = (uint)((DateTimeOffset)DateTime.UtcNow).ToUnixTimeSeconds();
                }

                uint length = (uint)data.Length;

                // Write per-packet header (16 bytes)
                bw.Write(tsSec);        // ts_sec
                bw.Write(tsUsec);       // ts_usec
                bw.Write(length);       // incl_len
                bw.Write(length);       // orig_len
                
                // Write packet data
                bw.Write(data);
            }
            return true;
        }
        catch
        {
            return false;
        }
    }
    
    private static byte[] BuildEthernetFrame(PacketInfo pkt)
    {
        // 14-byte Ethernet header fallback
        // dst MAC: 00:11:22:33:44:55, src MAC: 66:77:88:99:aa:bb, EtherType: 0x0800 (IPv4)
        byte[] frame = new byte[14 + 20]; // 14 eth + 20 ipv4
        
        // Dst Mac
        frame[0] = 0x00; frame[1] = 0x11; frame[2] = 0x22; frame[3] = 0x33; frame[4] = 0x44; frame[5] = 0x55;
        // Src Mac
        frame[6] = 0x66; frame[7] = 0x77; frame[8] = 0x88; frame[9] = 0x99; frame[10] = 0xaa; frame[11] = 0xbb;
        // EtherType IPv4
        frame[12] = 0x08; frame[13] = 0x00;

        // Dummy payload
        for (int i = 14; i < frame.Length; i++)
        {
            frame[i] = 0x00;
        }
        
        return frame;
    }
}
